src/Icicle/Avalanche/Statement/Flatten/Save.hs

```diff
@@ -0,0 +1,150 @@
+    MapT _ _ -> no
+
+    -- A buffer! Look inside and mark them as necessary.
+    BufT n t 
+     | FactIdentifierT <- t
+     -> forBuf xx n t (yes . KeepFactInHistory)
+     -- Allow for nested buffers: while we don't allow them now, they may be useful in the future
+     | otherwise
+     -> forBuf xx n t (go t)
+
+    -- Look through the contents of the array for any Buf FactIdentifierT.
+    ArrayT t
+     -> forArr xx t (go t)
+
+    -- The Option may contain a nested Buf FactIdentifierT, so try it
+    OptionT t
+     -> do s <- go t (fpOptionGet t `xApp` xx)
+           case s of
+            -- If there is no nested buf, we must be very careful to return Nothing
+            -- Otherwise we might loop through this stuff for no reason.
+            Nothing -> no
+            Just s' -> yes $ If (fpIsSome t `xApp` xx) s' mempty
+
+    PairT t u
+     -> do s1 <- go t (fpFst t u `xApp` xx)
+           s2 <- go u (fpSnd t u `xApp` xx)
+           -- If either side contains something, this is something
+           return (s1 <> s2)
+
+    SumT t u
+     -> do s1 <- go t (fpSumLeft  t u `xApp` xx)
+           s2 <- go u (fpSumRight t u `xApp` xx)
+           case (s1 <> s2) of
+            Nothing
+             -> no
+            Just _ 
+             -> yes
+              $ If (fpIsRight t u `xApp` xx) (stmtOf s2) (stmtOf s1)
+
+    StructT st
+     -> mconcat <$> mapM (\(nm,t) -> go t (fpStructGet nm t st `xApp` xx))
+                    (Map.toList (getStructType st))
+
+
+
+ where
+  no = return Nothing
+  yes = return . Just
+
+  go t x' = flattenSave' a_fresh x' t
+
+  forBuf x n t f
+   = do nm' <- fresh
+        s <- forArr (xVar nm') t f
+        case s of
+         Nothing -> no
+         Just s' -> yes
+                  $ Let nm' (xPrim (Flat.PrimBuf $ Flat.PrimBufRead n t) `xApp` x) s'
+
+  forArr x t f
+   = do nm' <- fresh
+        let ix = xPrim (Flat.PrimUnsafe $ Flat.PrimUnsafeArrayIndex t) `makeApps'` [x, xVar nm']
+        f'  <- f ix
+        case f' of
+         Nothing -> no
+         Just s -> yes
+                 $ ForeachInts nm' xZero (xPrim (Flat.PrimProject $ Flat.PrimProjectArrayLength t) `xApp` x) s
+
+  stmtOf = fromMaybe mempty
+
+  -- Annotation plumbing
+  makeApps' = makeApps a_fresh
+  xVar      = XVar     a_fresh
+  xPrim     = XPrim    a_fresh
+  xValue    = XValue   a_fresh
+  xApp      = XApp     a_fresh
+
+  xZero     = xValue IntT (VInt 0)
+
+  fpIsSome    t = xPrim (Flat.PrimProject (Flat.PrimProjectOptionIsSome t))
+  fpOptionGet t = xPrim (Flat.PrimUnsafe (Flat.PrimUnsafeOptionGet t))
+
+  fpIsRight  t u = xPrim (Flat.PrimProject (Flat.PrimProjectSumIsRight t u))
+  fpSumLeft  t u = xPrim (Flat.PrimUnsafe (Flat.PrimUnsafeSumGetLeft  t u))
+  fpSumRight t u = xPrim (Flat.PrimUnsafe (Flat.PrimUnsafeSumGetRight t u))
+
+  fpFst t u     = xPrim (Flat.PrimMinimal $ Min.PrimPair $ Min.PrimPairFst t u)
+  fpSnd t u     = xPrim (Flat.PrimMinimal $ Min.PrimPair $ Min.PrimPairSnd t u)
```

Comment:
I've seen these functions elsewhere I think, perhaps it'd be good to make a `Compounds` module for them at some point?